### PR TITLE
feat: Support Elastic Beanstalk backwards compatibility via the CLI

### PR DESCRIPTION
**Pull request by 96malhar**
**Note** - The CI pipeline is not automatically run when trying to merge a PR into a staging branch. I manually triggered the aws-dotnet-deploy-ci CodeBuild job and it ran successfully (build 223 and 227).

*Issue #, if available:*
DOTNET-5510

*Description of changes:*
This PR supports the CLI workflow to deploy to an existing Elastic Beanstalk environment. **It does not update server mode**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/AWS.Deploy.Orchestration/DeploymentCommands/BeanstalkEnvironmentDeploymentCommand.cs [hunk 1]
```diff
@@ -3,19 +3,66 @@
...
+                throw new ElasticBeanstalkException(DeployToolErrorCode.FailedToUpdateElasticBeanstalkEnvironment, "Failed to update the Elastic Beanstalk environment");
+            }
+        }
+        public async Task<List<DisplayedResourceItem>> GetDeploymentOutputsAsync(IDisplayedResourcesHandler displayedResourcesHandler, CloudApplication cloudApplication, Recommendation recommendation)
+        {
+            var displayedResources = new List<DisplayedResourceItem>();
+            var environment = await displayedResourcesHandler.AwsResourceQueryer.DescribeElasticBeanstalkEnvironment(cloudApplication.Name);
+            var data = new Dictionary<string, string>() {{ "Endpoint", $"http://{environment.CNAME}/" }};
+            var description = "An AWS Elastic Beanstalk environment is a collection of AWS resources running an application version.";
+            displayedResources.Add(new DisplayedResourceItem(environment.EnvironmentName, description, "Elastic Beanstalk Environment", data));
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/DeploymentCommands/BeanstalkEnvironmentDeploymentCommand.cs:64]: Make the type parameter a constant. We should also sync with Chris to see if they need more information for the toolkit.

File: src/AWS.Deploy.Orchestration/ElasticBeanstalkHandler.cs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+using AWS.Deploy.Common.IO;
+
+namespace AWS.Deploy.Orchestration
+{
+    public interface IElasticBeanstalkHandler
+    {
+        Task<S3Location> CreateApplicationStorageLocationAsync(string applicationName, string versionLabel, string deploymentPackage);
+        Task<CreateApplicationVersionResponse> CreateApplicationVersionAsync(string applicationName, string versionLabel, S3Location sourceBundle);
+        Task<bool> UpdateEnvironmentAsync(string applicationName, string environmentName, string versionLabel, List<ConfigurationOptionSetting> optionSettings);
+        List<ConfigurationOptionSetting> GetAdditionalSettings(Recommendation recommendation);
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/ElasticBeanstalkHandler.cs:21]: What makes these settings "Additional"? Either come up with a better name or add docs to the method.

**Review by normj — APPROVED** [review]
Just a couple minor comments but looks good.

File: src/AWS.Deploy.CLI/Commands/DeployCommand.cs [hunk 3]
```diff
@@ -289,15 +298,15 @@ private async Task<Recommendation> GetSelectedRecommendationFromPreviousDeployme
...
 
-            // currently the ability to load and apply previous settings is available only if the deployment target is a CloudFormation Stack
-            // Do not have an easy way to capture previous settings without querying multiple APIs for non CloudFormation based deployments
-            if (deployedApplication.ResourceType != CloudApplicationResourceType.CloudFormationStack)
-                return selectedRecommendation;
+            IDictionary<string, object> previousSettings;
+            if (deployedApplication.ResourceType == CloudApplicationResourceType.CloudFormationStack)
+                previousSettings = (await _templateMetadataReader.LoadCloudApplicationMetadata(deployedApplication.Name)).Settings;
+            else
+                previousSettings = (await _deployedApplicationQueryer.GetPreviousSettings(deployedApplication));
```
Comment (philasmar) [inline src/AWS.Deploy.CLI/Commands/DeployCommand.cs:305]: `_deployedApplicationQueryer.GetPreviousSettings` is essentially the generic function to get previous settings. Can you move `previousSettings = (await _templateMetadataReader.LoadCloudApplicationMetadata(deployedApplication.Name)).Settings;` also into that method? I don't see why they need to be separated at this point.

File: src/AWS.Deploy.Common/Exceptions.cs [hunk 4]
```diff
@@ -91,7 +91,12 @@ public enum DeployToolErrorCode
         InvalidSaveDirectoryForCdkProject = 10006900,
         FailedToFindDeploymentProjectRecipeId = 10007000,
         UnexpectedError = 10007100,
-        FailedToCreateDeploymentCommandInstance = 10007200
+        FailedToCreateDeploymentCommandInstance = 10007200,
```
Comment (philasmar) [inline src/AWS.Deploy.Common/Exceptions.cs:94]: The PR that I already merged in has an error code 10007200. Please increment all the new ones.

File: src/AWS.Deploy.Common/IO/FileManager.cs [hunk 5]
```diff
@@ -31,6 +34,12 @@ public class FileManager : IFileManager
         public Task WriteAllTextAsync(string filePath, string contents, CancellationToken cancellationToken) =>
             File.WriteAllTextAsync(filePath, contents, cancellationToken);
 
+        public FileStream OpenRead(string filePath) => File.OpenRead(filePath);
+
+        public string GetExtension(string filePath) => Path.GetExtension(filePath);
+
+        public long GetSizeInBytes(string filePath) => new FileInfo(filePath).Length;
```
Comment (philasmar) [inline src/AWS.Deploy.Common/IO/FileManager.cs:41]: I'm worried about null pointer exceptions here. Can you do some sanity checks here?

File: src/AWS.Deploy.Constants/ElasticBeanstalk.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        public const string HealthCheckURLOptionNameSpace = "aws:elasticbeanstalk:application";
+        public const string HealthCheckURLOptionName = "Application Healthcheck URL";
+
+        public static List<Tuple<string, string, string>> OptionSettingQueryList = new()
+        {
+            new (EnhancedHealthReportingOptionId, EnhancedHealthReportingOptionNameSpace, EnhancedHealthReportingOptionName),
+            new (XRayTracingOptionId, XRayTracingOptionNameSpace, XRayTracingOptionName),
+            new (ProxyOptionId, ProxyOptionNameSpace, ProxyOptionName),
+            new (HealthCheckURLOptionId, HealthCheckURLOptionNameSpace, HealthCheckURLOptionName)
+        };
```
Comment (philasmar) [inline src/AWS.Deploy.Constants/ElasticBeanstalk.cs:31]: I'd prefer if we don't use a Tuple. They are very vague to work with. Either replace this by a class or add some documentation to explain what the different values are.

File: src/AWS.Deploy.Orchestration/DeploymentCommands/BeanstalkEnvironmentDeploymentCommand.cs [hunk 7]
```diff
@@ -3,19 +3,66 @@
...
+        {
+            if (orchestrator._interactiveService == null)
+                throw new InvalidOperationException($"{nameof(orchestrator._interactiveService)} is null as part of the orchestartor object");
+            if (orchestrator._awsResourceQueryer == null)
+                throw new InvalidOperationException($"{nameof(orchestrator._awsResourceQueryer)} is null as part of the orchestartor object");
+            if (orchestrator._elasticBeanstalkHandler == null)
+                throw new InvalidOperationException($"{nameof(orchestrator._elasticBeanstalkHandler)} is null as part of the orchestartor object");
+            if (orchestrator._s3Handler == null)
+                throw new InvalidOperationException($"{nameof(orchestrator._s3Handler)} is null as part of the orchestartor object");
+
```
Comment (philasmar) [inline src/AWS.Deploy.Orchestration/DeploymentCommands/BeanstalkEnvironmentDeploymentCommand.cs:25]: why are we accessing a lot of the services through the orchestrator and not using dependency injection on this class and accessing them directly?

File: src/AWS.Deploy.Orchestration/DeploymentCommands/BeanstalkEnvironmentDeploymentCommand.cs [hunk 8]
```diff
@@ -3,19 +3,66 @@
...
+            if (string.IsNullOrEmpty(applicationName))
+            {
+                var message = $"Could not find any Elastic Beanstalk application that contains the following environment: {environmentName}";
+                throw new AWSResourceNotFoundException(DeployToolErrorCode.FailedToFindElasticBeanstalkApplication, message);
+            }
+
+            orchestrator._interactiveService.LogMessageLine($"Initiating deployment to {cloudApplication.DisplayName}...");
+
+            var versionLabel = $"v-{DateTime.Now.Ticks}";
```
Comment (philasmar) [inline src/AWS.Deploy.Orchestration/DeploymentCommands/BeanstalkEnvironmentDeploymentCommand.cs:41]: nit: could you remove the dash between the "v" and ticks? versions are usually v1, v2, v3 and not v-1, v-2, v-3

File: src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs [hunk 9]
```diff
@@ -204,10 +225,37 @@ private async Task<List<CloudApplication>> GetExistingBeanstalkEnvironments()
...
+                optionSettings[optionId] = configurationSetting.Value;
+            }
+
+            return optionSettings;
+        }
+
+        private ConfigurationOptionSetting GetBeanstalkEnvironmentConfigurationSetting(List<ConfigurationOptionSetting> configurationSettings, string optionNameSpace, string optionName)
+        {
+            var configurationSetting = configurationSettings
+                .Where(x => string.Equals(optionNameSpace, x.Namespace) && string.Equals(optionName, x.OptionName))
```
Comment (philasmar) [inline src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs:255]: nit: you don't need to use Where. You can put the condition inside FirstOrDefault

**Review by philasmar — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.Orchestration/DeploymentCommands/BeanstalkEnvironmentDeploymentCommand.cs (hunk 7 above)
Comment (96malhar) [reply src/AWS.Deploy.Orchestration/DeploymentCommands/BeanstalkEnvironmentDeploymentCommand.cs:25]: All classes them implement the `IDeploymentCommand` interface are instantiated during runtime via the [DeploymentCommandFactory](https://github.com/aws/aws-dotnet-deploy/blob/kmalhar/beanstalk-backwards-compatibility/src/AWS.Deploy.Orchestration/DeploymentCommands/DeploymentCommandFactory.cs#L28). It is instantiated with a parameterless constructor and thus these classes cannot have any dependencies.

I agree that attaching each service handler in the `Orchestrator` will make it bloated. To handle this I created a single `AWSServiceHander` class which will be part of the `Orchestrator` and all individual service handlers will be a part of this class via dependency injection.

**Comment by 96malhar** [comment]
> Also I noticed that this is not being merged to dev. Is that intentional?

Yes, this is intentional because I am merging my changes in a staging branch. I have also mentioned this in the PR description.

**Review by philasmar — APPROVED** [review]

**Pull request merged**